core/src/forwarding_stage.rs

```diff
@@ -45,95 +42,203 @@ mod packet_container;
 /// this should be evaluated with new stage.
 const FORWARD_BATCH_SIZE: usize = 128;
 
-/// Addresses to use for forwarding.
-#[derive(Clone)]
-pub struct ForwardingAddresses {
-    /// Forwarding address for TPU (non-votes)
-    pub tpu: Option<SocketAddr>,
-    /// Forwarding address for votes
-    pub tpu_vote: Option<SocketAddr>,
-}
+/// How far ahead to look in the leader schedule when determining forwarding
+/// addresses. The unit is `NUM_CONSECUTIVE_LEADER_SLOTS`.
+///
+/// This lookahead is needed because the immediate next leader might not have
+/// shared their forwarding ports. In such cases, we skip them and attempt to
+/// forward to the next available leader (up to this limit).
+const NUM_LOOKAHEAD_LEADERS: u64 = 2;
```

Comment:
Maybe we should make this 3 so we reduce the probability of malicious leaders acting funny and removing their TPU port info just to mess with forwarder impl?

Reply:
@alexpyattaev not really understand why 3 is different from 2. I should admit that has been chosen just because it seems to be better than 1. But not sure if use 3 or 4.
I forgot to say in comment that the unit is NUN_CONSECUTIVE_SLOTS which is 4. So when I say here 2, it means that I will check next 8 slots.

Reply:
yes that is clear. more leaders = harder for one entity to control them all, creating pathological cases where we do not have anywhere to forward.